### PR TITLE
Enable editable scenarios

### PR DESCRIPTION
**Pull request by bugsz**
The same as #6, but in a new branch name.
The first commit addresses #3, and the following commits address #4 

Btw whether the scenario is editable does not relate to the omniscient? (i.e. we can still only show partial information while making the scenario editable?)

### CODE REVIEW

**Comment by XuhuiZhou** [comment]
Good point! This is just a feature that I am suggesting. Cuz once people want to edit things, that usually implies they want full control. This is basically making things easier to handle.

**Review by XuhuiZhou — CHANGES_REQUESTED** [review]
image
Some meta-comments:
* Instead of listing the "create your own scenario" on the side bar. Create a button on the above to with text like "editable", once it is selected, all the things become editable and goals become omniscient
* Put scenario in a single row, besides just listing the codename, include a few tokens of the scenario, so people could have a preview of the scenario before choing them
* Allow choose models for each agent, include a special human model, which would require human input. If all models are not humans, then we do normal sotopia simulation.

**Comment by XuhuiZhou** [comment]
@bugsz thanks for the update, some new feedback
Screenshot 2024-08-14 at 12 14 49 PM

**Comment by bugsz** [comment]
Still in progress..
Currently support the flexible choice of agents (including human-human and model-model interaction)

**Comment by bugsz** [comment]
A bit strange behavior on my end:
Now I can edit the scenario and agent bg. I can see the change reflected in the environment (and also in LLMAgent.inbox). But I did not see the conversation change as expected (for example, I make one agent to be `giselle does not want to talk` but the conversation is still long. @XuhuiZhou Is it possible for you to see what's happening?

**Comment by XuhuiZhou** [comment]
> A bit strange behavior on my end: Now I can edit the scenario and agent bg. I can see the change reflected in the environment (and also in LLMAgent.inbox). But I did not see the conversation change as expected (for example, I make one agent to be `giselle does not want to talk` but the conversation is still long. @XuhuiZhou Is it possible for you to see what's happening?
Try debugging by outputting the agent prompt?

File: socialstream/chat/callbacks.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    print("Edited agent 2: ", agent_2_profile)
+
+    env_agent_combo = EnvAgentProfileCombo(
+        env=st.session_state.env.profile,
+        agents=[agent_1_profile, agent_2_profile],
+    )
+    set_from_env_agent_profile_combo(env_agent_combo=env_agent_combo, reset_msgs=False)
+
+
+from sotopia.database import EpisodeLog
```
Comment (XuhuiZhou) [inline socialstream/chat/callbacks.py:121]: move this to the top

File: socialstream/chat/callbacks.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        ],
+        messages=[
+            [(m[0], m[1], m[2].to_natural_language()) for m in messages_in_turn]
+            for messages_in_turn in messages
+        ],
+        reasoning=reasoning,
+        rewards=rewards,
+        rewards_prompt="",
+    )
+    from socialstream.rendering_utils import render_messages
```
Comment (XuhuiZhou) [inline socialstream/chat/callbacks.py:148]: move this?

File: socialstream/chat/callbacks.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            message_list.append(
+                f"{message['role']} {message['type']}: {message['content']}"
+            )
+        else:
+            message_list.append(f"{message['content']}")
+
+    message_list = [message.replace("**", "") for message in message_list]
+    return "\n".join(message_list)
+
+    # print(epilog.render_for_humans())
```
Comment (XuhuiZhou) [inline socialstream/chat/callbacks.py:169]: remove this

File: socialstream/rendering_utils.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+
+def get_secret_info(profile: AgentProfile, display_name: bool = True) -> str:
+    if profile.age == 0:
+        return profile.secret
+    else:
+        return agent_profile_to_secret_info(profile, display_name=display_name)
+
+
+def _agent_profile_to_friendabove_self(
```
Comment (XuhuiZhou) [inline socialstream/rendering_utils.py:240]: why we are re-writing this?

File: socialstream/rendering_utils.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+                agent.profile, agent_id, display_name=False
+            ),
+            agent_id=agent_id,
+        )
+        for agent_id, agent in enumerate(agents.values())
+    ]
+    return agent_to_render
+
+
+def render_messages(
```
Comment (XuhuiZhou) [inline socialstream/rendering_utils.py:183]: shouldn't this be in the rendering folder?

File: socialstream/chat/chat_omniscient.py
Comment (XuhuiZhou) [inline socialstream/chat/chat_omniscient.py:1]: qq, isn't chat_omniscient kind of replicating a lot of features in chat.py? Can we combine them?

**Review by XuhuiZhou — APPROVED** [review]
Thanks for the efforts, I left some comments but for the sake of moving fast. I will let you decide what to merge.

**Comment by XuhuiZhou** [comment]
Also, I notice that we use the callback functions. Any chance you can update the readme to introduce/explain that concept?
@bugsz

**Comment by bugsz** [comment]
@XuhuiZhou I think ready to merge!

**Pull request merged**